Windows: fix test failure - TestShardWriter_Write_ErrDialTimeout

mvadu opened the pull request:
Windows does not timeout when the timeout is set to 1Nano sec. This causes TestShardWriter_Write_ErrDialTimeout to fail in Windows.

```go
=== RUN TestShardWriter_Write_ErrDialTimeout
[cluster] 2016/02/10 09:32:30 Starting cluster service
[cluster] 2016/02/10 09:32:30 accept remote connection from 127.0.0.1:57034
[cluster] 2016/02/10 09:32:30 unable to read type-length-value read message type: WSARecv tcp 127.0.0.1:57033: use of closed network connection
[cluster] 2016/02/10 09:32:30 close remote connection from 127.0.0.1:57034
[cluster] 2016/02/10 09:32:30 cluster service accept error: network connection closed
--- FAIL: TestShardWriter_Write_ErrDialTimeout (0.00s)
	shard_writer_test.go:162: expected error <nil>, to contain i/o timeout
```

As a fix I added 1 nano sec sleep in WriteShard, hopefully it won't cause any major performance impacts in Linux environments. Otherwise we will have to go for platform specific patch.

e-dard commented:
@mvadu this doesn't seem like a good way to fix this issue. If the test is consistently failing on Windows (I assume because of the Go runtime's time precision on that platform?), then the way to fix this test is to rewrite the test, such that it introduces an artificial delay to the Dial.

Does that make sense?

mvadu commented:
@e-dard agreed. But the problem here is timeout is passed as part of constructor, and used in `WriteShard` routine. There is no way to introduce the delay to the Dial from the test routine. 

Shall I make the delay a conditional one, to add 1ns delay only if timeout duration is 1Nanosec? but then the cost of evaluating the condition may be longer than 1NS. Suggestions?

jwilder commented:
@mvadu Adding a sleep isn't the right fix here.  Agree w/ @e-dard.  Probably need to restructure the code to be able to force Dial to return a timeout error.

mvadu commented:
@e-dard , @jwilder is there any reason why we use `io.Writer` interface in `WriteTLV`, where as the connection is actually a `net.Conn`? I am seeing many instances in golang google group about `SetWriteDeadline` not timing out properly. One of the theories given (e.g. https://groups.google.com/forum/#!topic/golang-nuts/L0igp3tax1w) is sending too small data packet which might not block for the deadline time, causing it not to throw timeout.

I explored that, and added prints to get return value of `w.Write` in WriteTLV. 

```go
=== RUN TestShardWriter_Write_ErrDialTimeout
[cluster] 2016/02/11 07:19:12 Starting cluster service
1
54
[cluster] 2016/02/11 07:19:12 accept remote connection from 127.0.0.1:52797
1
2
[cluster] 2016/02/11 07:19:12 unable to read type-length-value read message type: WSARecv tcp 127.0.0.1:52796: use of closed network connection
[cluster] 2016/02/11 07:19:12 close remote connection from 127.0.0.1:52797
[cluster] 2016/02/11 07:19:12 cluster service accept error: network connection closed
--- FAIL: TestShardWriter_Write_ErrDialTimeout (0.00s)
	shard_writer_test.go:166: expected error <nil>, to contain i/o timeout
```

As you can see actual io.Writer interface implementation did not throw any error here, it returned actual bytes being written to. As per documentation "Write must return a non-nil error if it returns n < len(p).". I am assuming it wrote all 54 bytes without any error. I changed the test to have 1000 points, and got around 52k bytes written to socket without any errors.

I removed the delay, and made a change to WriteTLV. Can you please try running a test on Linux where we know the test works, and share the results?

mvadu commented:
@jwilder @e-dard I think the issue is with how net.DialTimeout (used at [#153](https://github.com/influxdata/influxdb/blob/8e100998fab81a70f8b8fa8fbb41e1ed8cafa6f5/cluster/shard_writer.go#L153)) is implemented in Windows. I got lost in [dial source](https://github.com/golang/go/blob/master/src/net/dial.go). Finally the fix seems to be using 0 which basically treats it as if no timeout, and it works on both Windows and Linux (verified on Raspbian Jessie).

e-dard commented on cluster/shard_writer_test.go:
```diff
@@ -147,13 +147,15 @@ func TestShardWriter_Write_ErrDialTimeout(t *testing.T) {
 
-	w := cluster.NewShardWriter(time.Nanosecond, 1)
+	//Windows implementation of net DialTimeout seems to fail with 1 Nano sec timeouts. Without a timeout
+	w := cluster.NewShardWriter(time.Nanosecond*0, 1)
```
@mvadu  can you tidy this up a bit please?

```go
// zero timeout set to support all platforms.
w := cluster.NewShardWriter(0, 1)
```

mvadu commented:
@e-dard done!! Sorry for the mess before.

e-dard commented:
LGTM :+1: 

/cc @jwilder

jwilder commented:
:+1:

mvadu commented:
Thanks!
